sretoolbox/container/image.py

```diff
@@ -57,14 +57,31 @@ class Image:
     :param password: (optional) The private registry password
     :param auth_server: (optional) The host that the username and password are
                         meant for
+
+    :param response_cache (optional) A cache to store responses,
+    indexed by both etag and digest
     """
+
+    ACCEPT_HEADERS = {
+        'Accept':
+        'application/vnd.docker.distribution.manifest.v1+json,'
+        'application/vnd.docker.distribution.manifest.v2+json,'
+        'application/vnd.docker.distribution.manifest.v1+prettyjws,'
+    }
+
+    MAX_CACHE_ITEM_SIZE = 50*1024
+
     def __init__(self, url, tag_override=None, username=None, password=None,
-                 auth_server=None):
+                 auth_server=None, response_cache=None):
         image_data = self._parse_image_url(url)
         self.scheme = image_data['scheme']
         self.registry = image_data['registry']
         self.repository = image_data['repository']
         self.image = image_data['image']
+        if response_cache is None:
+            self.response_cache = {}
```

Comment:
If a cache backend is bot provided, we should not cache at all. The cache is only useful when shared between instances

Reply:
Yes. But as I mention in the description the code is way simpler if I fake it locally and keep a one-entry cache, instead of polluting the consumers with awful `if self.response_cache` blocks.

Reply:
`if` is cheaper than `request.head`. Way chepaer.

Reply:
1. Performance doesn't seem to be a problem in here. GET request quota is.
2. Show me a way around `head` without conditional requests, which aren't implemented in our most pressing use case.

Reply:
1. Performance doesn't seem to be a problem in here. GET request quota is.

You're making a lot of useless requests. To solve your use-case, you're penalizing whoever else is using this lib, when you don't really have to. This is a community project, we should assume there are more users than qontract-reconcile. Even internally, many other projects are using this lib.

2. Show me a way around head without conditional requests, which aren't implemented in our most pressing use case.

```
if self.response_cache is None:
```